prometheus/reporter.go

```diff
@@ -0,0 +1,254 @@
+	ctr := &prom.CounterVec{}
+	id := hashMetricLabelsToID(name, tags)
+	exists := r.hasCounter(id)
+	if exists {
+		return ctr, errorAlreadyRegistered
+	}
+	labelKeys := keysFromMap(tags)
+	ctr = prom.NewCounterVec(
+		prom.CounterOpts{
+			Name: name,
+			Help: desc,
+		},
+		labelKeys,
+	)
+	err := prom.Register(ctr)
+	if err != nil {
+		return ctr, err
+	}
+	r.mtx.Lock()
+	defer r.mtx.Unlock()
+	r.counters[id] = ctr
+	return ctr, nil
+}
+
+// ReportCounter reports a counter value
+func (r *Reporter) ReportCounter(name string, tags map[string]string, value int64) {
+	id := hashMetricLabelsToID(name, tags)
+
+	r.mtx.RLock()
+	ctr, ok := r.counters[id]
+	r.mtx.RUnlock()
+
+	if !ok {
+		var err error
+		ctr, err = r.RegisterCounter(name, tags, name+" counter")
+		if err != nil {
+			panic(err)
+		}
+	}
+	ctr.With(tags).Add(float64(value))
+}
+
+// RegisterGauge is a helper method to initialize a gauge in the prometheus backend with a given help text.
+// If not called explicitly, the Reporter will create one for you on first use, with a not super helpful HELP string
+func (r *Reporter) RegisterGauge(name string, tags map[string]string, desc string) (*prom.GaugeVec, error) {
+	g := &prom.GaugeVec{}
+	id := hashMetricLabelsToID(name, tags)
+	exists := r.hasGauge(id)
+	if exists {
+		return g, errorAlreadyRegistered
+	}
+	labelKeys := keysFromMap(tags)
+
+	g = prom.NewGaugeVec(
+		prom.GaugeOpts{
+			Name: name,
+			Help: desc,
+		},
+		labelKeys,
+	)
+	err := prom.Register(g)
+	if err != nil {
+		return g, err
+	}
+	r.mtx.Lock()
+	defer r.mtx.Unlock()
+	r.gauges[id] = g
+	return g, nil
+}
+
+// ReportGauge reports a gauge value
+func (r *Reporter) ReportGauge(name string, tags map[string]string, value float64) {
+	id := hashMetricLabelsToID(name, tags)
+
+	r.mtx.RLock()
+	g, ok := r.gauges[id]
+	r.mtx.RUnlock()
+
+	if !ok {
+		var err error
+		g, err = r.RegisterGauge(name, tags, name+" gauge")
+		if err != nil {
+			panic(err)
+		}
+	}
+	g.With(tags).Set(value)
+}
+
+// RegisterTimer is a helper method to initialize a Timer histogram vector in the prometheus backend with a given help text.
+// If not called explicitly, the Reporter will create one for you on first use, with a not super helpful HELP string
+// objectives is the prometheus.SummaryVec objectives. Default, if nil passed is {0.5: 0.05, 0.9: 0.01, 0.99: 0.001, 0.999: 0.0001}
+// See https://godoc.org/github.com/prometheus/client_golang/prometheus#SummaryOpts
+func (r *Reporter) RegisterTimer(name string, tags map[string]string, desc string, objectives map[float64]float64) (*prom.SummaryVec, error) {
+	h := &prom.SummaryVec{}
+	id := hashMetricLabelsToID(name, tags)
+	exists := r.hasSummary(id)
+	if exists {
+		return h, errorAlreadyRegistered
+	}
+	labelKeys := keysFromMap(tags)
+
+	if objectives == nil {
+		objectives = map[float64]float64{0.5: 0.05, 0.9: 0.01, 0.99: 0.001, 0.999: 0.0001}
```

Comment:
Perhaps make the histogram defaults an option of the reporter with a single default defined on creation if none specified at all here?

That way users can register timers whatever their desired default histograms are just by passing `nil` here.